src/kernel/src/fs/file.rs

```diff
@@ -67,13 +68,104 @@ pub trait VFileOps: Debug + Send + Sync + Display {
     fn ioctl(
         &self,
         file: &VFile,
-        com: u64,
+        com: IoctlCom,
         data: &mut [u8],
         cred: &Ucred,
         td: &VThread,
     ) -> Result<(), Box<dyn Errno>>;
 }
 
+#[repr(transparent)]
+#[derive(Debug, Clone, Copy, Eq, PartialEq)]
+pub struct IoctlCom(u32);
+
+impl Display for IoctlCom {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(f, "0x{:x}", self.0)
+    }
+}
+
+impl IoctlCom {
+    pub const IOCPARM_SHIFT: u32 = 13;
+    pub const IOCPARM_MASK: u32 = (1 << Self::IOCPARM_SHIFT) - 1;
+    pub const IOC_VOID: u32 = 0x20000000;
+    pub const IOC_OUT: u32 = 0x40000000;
+    pub const IOC_IN: u32 = 0x80000000;
+    pub const IOC_INOUT: u32 = Self::IOC_IN | Self::IOC_OUT;
+
+    pub const fn try_from_raw(com: u64) -> Result<Self, SysErr> {
+        let com = com as u32;
+
+        if Self::is_invalid(com) {
+            return Err(SysErr::Raw(ENOTTY));
+        }
+
+        Ok(Self(com))
+    }
+
+    pub const fn new(inout: u32, group: u8, num: u8, len: usize) -> Self {
+        let len: u32 = if len > (u32::MAX) as usize {
+            panic!("IOCPARM_LEN is too large");
+        } else {
+            len as u32
+        };
+
+        Self(inout | ((len & Self::IOCPARM_MASK) << 16) | ((group as u32) << 8) | (num as u32))
+    }
+
+    pub fn size(&self) -> usize {
+        Self::iocparm_len(self.0)
+    }
+
+    pub fn is_void(&self) -> bool {
+        self.0 & Self::IOC_VOID != 0
+    }
+
+    pub fn is_out(&self) -> bool {
+        self.0 & Self::IOC_OUT != 0
+    }
+
+    pub fn is_in(&self) -> bool {
+        self.0 & Self::IOC_IN != 0
+    }
+
+    pub const fn is_invalid(com: u32) -> bool {
+        (com & (Self::IOC_VOID | Self::IOC_IN | Self::IOC_OUT) == 0)
+            || (com & (Self::IOC_IN | Self::IOC_OUT)) != 0 && Self::iocparm_len(com) == 0
+            || (com & Self::IOC_VOID != 0
+                && Self::iocparm_len(com) != 0
+                && Self::iocparm_len(com) != 4)
```

Comment:
Is it possible to split the whole this into a separated `if` so it will be more easier to read?